core/services/relay/evm/functions/logpoller_wrapper.go

```diff
@@ -55,29 +98,201 @@ func (l *logPollerWrapper) Ready() error {
+		_ = subscriber.UpdateRoutes(l.routerContract.Address(), l.routerContract.Address())
 	}
 }
+
+func (l *logPollerWrapper) checkForRouteUpdates() {
+	defer l.closeWait.Done()
+	freqSec := l.pluginConfig.ContractUpdateCheckFrequencySec
+	if freqSec == 0 {
+		l.lggr.Errorw("ContractUpdateCheckFrequencySec is zero - route update checks disabled")
+		return
+	}
+
+	updateOnce := func() {
+		// NOTE: timeout == frequency here, could be changed to a separate config value
+		timeoutCtx, cancel := utils.ContextFromChanWithTimeout(l.stopCh, time.Duration(l.pluginConfig.ContractUpdateCheckFrequencySec)*time.Second)
+		defer cancel()
+		active, proposed, err := l.getCurrentCoordinators(timeoutCtx)
+		if err != nil {
+			l.lggr.Errorw("LogPollerWrapper: error calling getCurrentCoordinators", "err", err)
+		}
+		l.handleRouteUpdate(active, proposed)
+	}
+
+	updateOnce() // update once right away
+	ticker := time.NewTicker(time.Duration(freqSec) * time.Second)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-l.stopCh:
+			return
+		case <-ticker.C:
+			updateOnce()
+		}
+	}
+}
+
+func (l *logPollerWrapper) getCurrentCoordinators(ctx context.Context) (common.Address, common.Address, error) {
+	if l.pluginConfig.ContractVersion == 0 {
+		return l.routerContract.Address(), l.routerContract.Address(), nil
+	}
+	var donId [32]byte
+	copy(donId[:], []byte(l.pluginConfig.DONId))
+
+	activeCoordinator, err := l.routerContract.GetContractById(&bind.CallOpts{
+		Pending: false,
+		Context: ctx,
+	}, donId, false)
+	if err != nil {
+		return common.Address{}, common.Address{}, err
+	}
+
+	proposedCoordinator, err := l.routerContract.GetContractById(&bind.CallOpts{
+		Pending: false,
+		Context: ctx,
+	}, donId, true)
+	if err != nil {
+		return common.Address{}, common.Address{}, err
+	}
+
+	return activeCoordinator, proposedCoordinator, nil
+}
+
+func (l *logPollerWrapper) handleRouteUpdate(activeCoordinator common.Address, proposedCoordinator common.Address) {
+	changed := false
+	l.mu.Lock()
+	if activeCoordinator != l.activeCoordinator || proposedCoordinator != l.proposedCoordinator {
+		l.activeCoordinator = activeCoordinator
+		l.proposedCoordinator = proposedCoordinator
+		changed = true
+	}
+	l.mu.Unlock()
+	if !changed {
+		l.lggr.Debug("LogPollerWrapper: no changes to routes")
+		return
+	}
+
+	l.lggr.Debugw("LogPollerWrapper: new routes", "activeCoordinator", activeCoordinator.Hex(), "proposedCoordinator", proposedCoordinator.Hex())
+	l.registerFilters(activeCoordinator)
+	l.registerFilters(proposedCoordinator)
+
+	for _, subscriber := range l.subscribers {
+		err := subscriber.UpdateRoutes(activeCoordinator, proposedCoordinator)
+		if err != nil {
+			l.lggr.Errorw("LogPollerWrapper: Failed to update routes", "error", err)
+		}
+	}
+}
+
+func filterName(addr common.Address) string {
+	return logpoller.FilterName("FunctionsLogPollerWrapper", addr.String())
+}
+
+func (l *logPollerWrapper) registerFilters(coordinatorAddress common.Address) error {
```

Comment:
Instead of having each component, such as [the configPoller](https://github.com/smartcontractkit/chainlink/pull/9978/files#diff-98105c6dc5340a154d3aae29f2f7d950d7cfeb1905d2844ac377546d17e330e1R185) or [the contractTransmitter](https://github.com/smartcontractkit/chainlink/pull/9978/files#diff-d4c4100c4ecc5bf07a91c701209d98ef7db89b0f1badbf82511b4699419a06acR241) handle filter registration, can we just handle all filter registration here?

Not sure what makes the most sense, but IMO either the logPollerWrapper should handle registration of all necessary filters, or each component should handle registration of the relevant filters, but not both.

Reply:
You're right, it could've been a little cleaner.
However, the logical distinction here is that filters need to be abstracted away from anything outside the relayer components (in our case Listener, which should be chain-agnostic).
Inside the relayer, all components are free to use logpoller directly because they are all EVM-aware.
I'll put a note to try and refactor this to put all filters in one place but I hope you're OK with this approach for now.

Reply:
That does make sense to me.  I'll approve as-is for now.